pkg/client/client_v2.go

```
@@ -0,0 +1,541 @@
+
+			switch action.Type {
+			case proto.ActionRequest_CUSTOM:
+				c.tryPerformAction(actionResults, action)
+			case proto.ActionRequest_DIAGNOSTICS:
+				// TODO: Implement the diagnostics action.
+				// At the moment it just returns action type unknown until implemented.
+				actionResults <- &proto.ActionResponse{
+					Token:  c.token,
+					Id:     action.Id,
+					Status: proto.ActionResponse_FAILED,
+					Result: ActionTypeUnknown,
+				}
+			default:
+				actionResults <- &proto.ActionResponse{
+					Token:  c.token,
+					Id:     action.Id,
+					Status: proto.ActionResponse_FAILED,
+					Result: ActionTypeUnknown,
+				}
+			}
+		}
+	}()
+
+	// action responses
+	actionsWrite.Add(1)
+	go func() {
+		defer actionsWrite.Done()
+
+		// initial connection of stream must send the token so
+		// the Elastic Agent knows this clients token.
+		err := actionsClient.Send(&proto.ActionResponse{
+			Token:  c.token,
+			Id:     ActionResponseInitID,
+			Status: proto.ActionResponse_SUCCESS,
+			Result: []byte("{}"),
+		})
+		if err != nil {
+			c.errChan <- err
+			return
+		}
+
+		for {
+			select {
+			case <-done:
+				return
+			case res := <-actionResults:
+				err := actionsClient.Send(res)
+				if err != nil {
+					// failed to send, add back to response to try again
+					actionResults <- res
+					c.errChan <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// wait for write goroutine to quit then close the client
+	actionsWrite.Wait()
+	actionsClient.CloseSend()
+
+	// then wait for the read to quit
+	actionsRead.Wait()
+}
+
+func (c *clientV2) tryPerformAction(actionResults chan *proto.ActionResponse, action *proto.ActionRequest) {
+	// find the unit
+	c.unitsLock.RLock()
+	unit := c.findUnit(action.UnitId, UnitType(action.UnitType))
+	c.unitsLock.RUnlock()
+	if unit == nil {
+		actionResults <- &proto.ActionResponse{
+			Token:  c.token,
+			Id:     action.Id,
+			Status: proto.ActionResponse_FAILED,
+			Result: ActionErrUnitNotFound,
+		}
+		return
+	}
```

So this mean that If an agent receive an `osquery` and it receive an osquery action and the unit isn't runnning we would silently ignore the event?